pkg/operator/controller/ttl_report.go

```diff
@@ -0,0 +1,95 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aquasecurity/starboard/pkg/apis/aquasecurity/v1alpha1"
+	"github.com/aquasecurity/starboard/pkg/operator/etc"
+	"github.com/aquasecurity/starboard/pkg/operator/predicate"
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type TTLReportReconciler struct {
+	logr.Logger
+	etc.Config
+	client.Client
+}
+
+func (r *TTLReportReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	installModePredicate, err := predicate.InstallModePredicate(r.Config)
+	if err != nil {
+		return err
+	}
+
+	err = ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.VulnerabilityReport{}, builder.WithPredicates(
+			predicate.Not(predicate.IsBeingTerminated),
+			installModePredicate)).
+		Complete(r.reconcileReport())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *TTLReportReconciler) reconcileReport() reconcile.Func {
+	return func(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+		log := r.Logger.WithValues("report", req.NamespacedName)
+
+		report := &v1alpha1.VulnerabilityReport{}
+		err := r.Client.Get(ctx, req.NamespacedName, report)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				log.V(1).Info("Ignoring cached report that must have been deleted")
+				return ctrl.Result{}, nil
+			}
+			return ctrl.Result{}, fmt.Errorf("getting report from cache: %w", err)
+		}
+
+		ttlReportAnnotationStr := report.Annotations[v1alpha1.TTLReportAnnotation]
+		if ttlReportAnnotationStr == "" {
+			return ctrl.Result{}, fmt.Errorf("the annotation %v is missing from the report", v1alpha1.TTLReportAnnotation)
+		}
+
+		creationTime := report.Report.UpdateTimestamp
+		ttlExpired, durationToTTLExpiration, err := ttlIsExpired(ttlReportAnnotationStr, creationTime.Time)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if ttlExpired {
+			log.V(1).Info("Removing vulnerabilityReport")
+			err := r.Client.Delete(ctx, report, &client.DeleteOptions{})
+			if err != nil && !errors.IsNotFound(err) {
+				return ctrl.Result{}, err
+			}
+			// Since the report is deleted there is no reason to requeue
+			return ctrl.Result{}, nil
+		}
+		log.V(1).Info("RequeueAfter", "durationToTTLExpiration", durationToTTLExpiration)
+		return ctrl.Result{RequeueAfter: durationToTTLExpiration}, nil
+	}
+}
+
+func ttlIsExpired(reportTTL string, creationTime time.Time) (bool, time.Duration, error) {
+	reportTTLTime, err := time.ParseDuration(reportTTL)
+	if err != nil {
+		return false, time.Duration(0), err
```

Comment:
Let's add more context to the error, e.g. `fmt.Errorf("failed parsing report TTL: %w")`. When we see it in the logs we'll know what went wrong without analysing stack traces.

Reply:
I moved this outside of the function since I use time.Duration instead of string in ttlIsExpired

When doing time.ParseDuration i check for error and return:
```.go
		reportTTLTime, err := time.ParseDuration(ttlReportAnnotationStr)
		if err != nil {
			return ctrl.Result{}, fmt.Errorf("failed parsing %v with value %v %w", v1alpha1.TTLReportAnnotation, ttlReportAnnotationStr, err)
		}
```